<?xml version="1.0"?>
<launch>
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find mixmcl)/config/custom_rosconsole.conf"/>
  <arg name="scan_topic" default="/p3dx/laser/scan"/>
  <arg name="global_localization" default="true"/>
  <arg name="init_x" default="4.0" />
  <arg name="init_y" default="8.4" />
  <arg name="init_a" default="3.14159265" />
  <arg name="max_particles" default="2000" />
  <arg name="resample_interval" default="1" />
  <arg name="launch-prefix" default="" />
  <arg name="record" default="true" />
  <arg name="bagfile" default="$(find mixmcl)/bags/ex3/2018-01-23-22-37-47-clean.bag"/>
  <arg name="map_yaml" default="$(find mixmcl)/maps/willowgarage/willowgarage-topleft.yaml" />
  <arg name="suicide"/>
  <!--
  <arg name="map_yaml" default="$(find p3dx_2dnav)/maps/willowgarage/willowgarage-refined.yaml" />
  -->
  <node pkg="mixmcl" type="mcmcl" name="mcmcl" output="screen" args="--run-from-bag $(arg bagfile)" launch-prefix="$(arg launch-prefix)">
    <remap from="scan" to="$(arg scan_topic)" />
    <rosparam command="load" file="$(find mixmcl)/yaml/amcl.yaml"/>
    <param unless="$(arg global_localization)" name="initial_pose_x" value="$(arg init_x)"/>
    <param unless="$(arg global_localization)" name="initial_pose_y" value="$(arg init_y)"/>
    <param unless="$(arg global_localization)" name="initial_pose_a" value="$(arg init_a)"/>
    <param name="global_localization" value="$(arg global_localization)"/>
    <param name="resample_interval" value="$(arg resample_interval)"/>
    <param name="max_particles" value="$(arg max_particles)"/>
    <param name="resample_type" value="kld"/>
    <!--
    -->
    <param name="dual_normalizer_ita" value="0.01"/>
    <param name="dual_loc_bandwidth" value="2"/>
    <param name="dual_ori_bandwidth" value="2"/>
    <param name="version1" value="false"/>
    <param name="static_update" value="false"/>
    <param name="demc_factor_gamma" value="2"/>
    <param name="demc_loc_bandwidth" value="0.01"/>
    <param name="demc_ori_bandwidth" value="0.01"/>
    <param name="laser_model_type" value="beam"/>
    <!--
    <param name="" value=""/>
    <param name="update_min_d" value="0.5"/>
    -->
  </node>

  <node if="$(arg record)" name="rosbagrecorder" pkg="rosbag" type="record" output="screen" args="-o mcmcl_mp$(arg max_particles)_ri$(arg resample_interval)_ita1e-2_gamma2 /tf /p3dx/laser/scan /p3dx/base_pose_ground_truth /mcl_pose /particlecloud /particlecloud2 /particlecloud3"/>
  <!--MAP SERVER-->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_yaml)" />
  <node if="$(arg suicide)" name="kill" pkg="mixmcl" type="kill.sh" args="/mcmcl" output="screen"/>

</launch>
